# Data Cleaning Basics

## 01. Reading CSV Files with Encodings

In [1]:
import numpy as np
import pandas as pd
laptops = pd.read_csv('laptops.csv', encoding='Latin-1')
laptops.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 13 columns):
Manufacturer                1303 non-null object
Model Name                  1303 non-null object
Category                    1303 non-null object
Screen Size                 1303 non-null object
Screen                      1303 non-null object
CPU                         1303 non-null object
RAM                         1303 non-null object
 Storage                    1303 non-null object
GPU                         1303 non-null object
Operating System            1303 non-null object
Operating System Version    1133 non-null object
Weight                      1303 non-null object
Price (Euros)               1303 non-null object
dtypes: object(13)
memory usage: 132.4+ KB


## 02. Cleaning Column Names

In [2]:
new_columns = []

for c in laptops.columns:
    clean_c = c.strip()
    new_columns.append(clean_c)

laptops.columns = new_columns

In [3]:
laptops

Manufacturer                                 Model Name  \
0           Apple                                MacBook Pro   
1           Apple                                Macbook Air   
2              HP                                     250 G6   
3           Apple                                MacBook Pro   
4           Apple                                MacBook Pro   
5            Acer                                   Aspire 3   
6           Apple                                MacBook Pro   
7           Apple                                Macbook Air   
8            Asus                            ZenBook UX430UN   
9            Acer                                    Swift 3   
10             HP                                     250 G6   
11             HP                                     250 G6   
12          Apple                                MacBook Pro   
13           Dell                              Inspiron 3567   
14          Apple                                MacBook 12"   
15          Apple                                MacBook Pro   
16           Dell                              Inspiron 3567   
17          Apple                                MacBook Pro   
18         Lenovo                          IdeaPad 320-15IKB   
19           Dell                                     XPS 13   
20           Asus                            Vivobook E200HA   
21         Lenovo                         Legion Y520-15IKBN   
22             HP                                     255 G6   
23           Dell                              Inspiron 5379   
24             HP    15-BS101nv (i7-8550U/8GB/256GB/FHD/W10)   
25           Dell                              Inspiron 3567   
26          Apple                                MacBook Air   
27           Dell                              Inspiron 5570   
28           Dell                              Latitude 5590   
29             HP                                ProBook 470   
...           ...                                        ...   
1273           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1274         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1275         Asus                       ZenBook UX305CA-UBM1   
1276         Dell                              Inspiron 3567   
1277         Acer                             Aspire ES1-531   
1278         Dell                              Inspiron 3552   
1279       Lenovo                         IdeaPad Y700-15ISK   
1280           HP                        Pavilion 15-AW003nv   
1281         Dell                              Inspiron 3567   
1282           HP                           Stream 11-Y000na   
1283         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1284       Lenovo                             Yoga 500-14ISK   
1285       Lenovo                             Yoga 900-13ISK   
1286       Lenovo                         IdeaPad 100S-14IBR   
1287           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1288         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1289         Asus                       ZenBook UX305CA-UBM1   
1290         Dell                              Inspiron 3567   
1291         Acer                             Aspire ES1-531   
1292         Dell                              Inspiron 3552   
1293       Lenovo                         IdeaPad Y700-15ISK   
1294           HP                        Pavilion 15-AW003nv   
1295         Dell                              Inspiron 3567   
1296           HP                           Stream 11-Y000na   
1297         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1298       Lenovo                             Yoga 500-14ISK   
1299       Lenovo                             Yoga 900-13ISK   
1300       Lenovo                         IdeaPad 100S-14IBR   
1301           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1302         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   

                Category Screen Size

## 03. Cleaning Column Names Continued

In [4]:
new_columns = []

def clean_col(col):
    col = col.strip()
    col = col.replace('Operating System', 'os')
    col = col.replace(' ', '_')
    col = col.replace('(', '')
    col = col.replace(')', '')
    col = col.lower()
    return col

for c in laptops.columns:
    clean_c = clean_col(c)
    new_columns.append(clean_c)

laptops.columns = new_columns

In [5]:
laptops

manufacturer                                 model_name  \
0           Apple                                MacBook Pro   
1           Apple                                Macbook Air   
2              HP                                     250 G6   
3           Apple                                MacBook Pro   
4           Apple                                MacBook Pro   
5            Acer                                   Aspire 3   
6           Apple                                MacBook Pro   
7           Apple                                Macbook Air   
8            Asus                            ZenBook UX430UN   
9            Acer                                    Swift 3   
10             HP                                     250 G6   
11             HP                                     250 G6   
12          Apple                                MacBook Pro   
13           Dell                              Inspiron 3567   
14          Apple                                MacBook 12"   
15          Apple                                MacBook Pro   
16           Dell                              Inspiron 3567   
17          Apple                                MacBook Pro   
18         Lenovo                          IdeaPad 320-15IKB   
19           Dell                                     XPS 13   
20           Asus                            Vivobook E200HA   
21         Lenovo                         Legion Y520-15IKBN   
22             HP                                     255 G6   
23           Dell                              Inspiron 5379   
24             HP    15-BS101nv (i7-8550U/8GB/256GB/FHD/W10)   
25           Dell                              Inspiron 3567   
26          Apple                                MacBook Air   
27           Dell                              Inspiron 5570   
28           Dell                              Latitude 5590   
29             HP                                ProBook 470   
...           ...                                        ...   
1273           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1274         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1275         Asus                       ZenBook UX305CA-UBM1   
1276         Dell                              Inspiron 3567   
1277         Acer                             Aspire ES1-531   
1278         Dell                              Inspiron 3552   
1279       Lenovo                         IdeaPad Y700-15ISK   
1280           HP                        Pavilion 15-AW003nv   
1281         Dell                              Inspiron 3567   
1282           HP                           Stream 11-Y000na   
1283         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1284       Lenovo                             Yoga 500-14ISK   
1285       Lenovo                             Yoga 900-13ISK   
1286       Lenovo                         IdeaPad 100S-14IBR   
1287           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1288         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1289         Asus                       ZenBook UX305CA-UBM1   
1290         Dell                              Inspiron 3567   
1291         Acer                             Aspire ES1-531   
1292         Dell                              Inspiron 3552   
1293       Lenovo                         IdeaPad Y700-15ISK   
1294           HP                        Pavilion 15-AW003nv   
1295         Dell                              Inspiron 3567   
1296           HP                           Stream 11-Y000na   
1297         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1298       Lenovo                             Yoga 500-14ISK   
1299       Lenovo                             Yoga 900-13ISK   
1300       Lenovo                         IdeaPad 100S-14IBR   
1301           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1302         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   

                category screen_size

## 04. Converting String Columns to Numeric

In [6]:
unique_ram = laptops['ram'].unique()

In [7]:
unique_ram

array(['8GB', '16GB', '4GB', '2GB', '12GB', '6GB', '32GB', '24GB', '64GB'],
      dtype=object)

## 05. Removing Non-Digit Characters

In [8]:
laptops['ram'] = laptops['ram'].str.replace('GB', '')
unique_ram = laptops['ram'].unique()

In [9]:
unique_ram

array(['8', '16', '4', '2', '12', '6', '32', '24', '64'], dtype=object)

## 06. Converting Columns to Numeric Dtypes

In [10]:
laptops["ram"] = laptops["ram"].str.replace('GB','')

laptops['ram'] = laptops['ram'].astype(int)
dtypes = laptops.dtypes

In [11]:
laptops

manufacturer                                 model_name  \
0           Apple                                MacBook Pro   
1           Apple                                Macbook Air   
2              HP                                     250 G6   
3           Apple                                MacBook Pro   
4           Apple                                MacBook Pro   
5            Acer                                   Aspire 3   
6           Apple                                MacBook Pro   
7           Apple                                Macbook Air   
8            Asus                            ZenBook UX430UN   
9            Acer                                    Swift 3   
10             HP                                     250 G6   
11             HP                                     250 G6   
12          Apple                                MacBook Pro   
13           Dell                              Inspiron 3567   
14          Apple                                MacBook 12"   
15          Apple                                MacBook Pro   
16           Dell                              Inspiron 3567   
17          Apple                                MacBook Pro   
18         Lenovo                          IdeaPad 320-15IKB   
19           Dell                                     XPS 13   
20           Asus                            Vivobook E200HA   
21         Lenovo                         Legion Y520-15IKBN   
22             HP                                     255 G6   
23           Dell                              Inspiron 5379   
24             HP    15-BS101nv (i7-8550U/8GB/256GB/FHD/W10)   
25           Dell                              Inspiron 3567   
26          Apple                                MacBook Air   
27           Dell                              Inspiron 5570   
28           Dell                              Latitude 5590   
29             HP                                ProBook 470   
...           ...                                        ...   
1273           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1274         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1275         Asus                       ZenBook UX305CA-UBM1   
1276         Dell                              Inspiron 3567   
1277         Acer                             Aspire ES1-531   
1278         Dell                              Inspiron 3552   
1279       Lenovo                         IdeaPad Y700-15ISK   
1280           HP                        Pavilion 15-AW003nv   
1281         Dell                              Inspiron 3567   
1282           HP                           Stream 11-Y000na   
1283         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1284       Lenovo                             Yoga 500-14ISK   
1285       Lenovo                             Yoga 900-13ISK   
1286       Lenovo                         IdeaPad 100S-14IBR   
1287           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1288         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1289         Asus                       ZenBook UX305CA-UBM1   
1290         Dell                              Inspiron 3567   
1291         Acer                             Aspire ES1-531   
1292         Dell                              Inspiron 3552   
1293       Lenovo                         IdeaPad Y700-15ISK   
1294           HP                        Pavilion 15-AW003nv   
1295         Dell                              Inspiron 3567   
1296           HP                           Stream 11-Y000na   
1297         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1298       Lenovo                             Yoga 500-14ISK   
1299       Lenovo                             Yoga 900-13ISK   
1300       Lenovo                         IdeaPad 100S-14IBR   
1301           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1302         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   

                category screen_size

In [12]:
dtypes

manufacturer    object
model_name      object
category        object
screen_size     object
screen          object
cpu             object
ram              int64
storage         object
gpu             object
os              object
os_version      object
weight          object
price_euros     object
dtype: object

## 07. Renaming Columns

In [13]:
laptops.rename({'ram': 'ram_gb'}, axis=1, inplace=True)
ram_gb_desc = laptops['ram_gb'].describe()

In [14]:
ram_gb_desc

count    1303.000000
mean        8.382195
std         5.084665
min         2.000000
25%         4.000000
50%         8.000000
75%         8.000000
max        64.000000
Name: ram_gb, dtype: float64

## 08. Extracting Values from Strings

In [15]:
laptops["gpu_manufacturer"] = (laptops["gpu"]
                                       .str.split()
                                       .str[0]
                              )

laptops['cpu_manufacturer'] = (laptops['cpu'].str.split().str[0])
cpu_manufacturer_counts = laptops['cpu_manufacturer'].value_counts()

In [16]:
cpu_manufacturer_counts

Intel      1240
AMD          62
Samsung       1
Name: cpu_manufacturer, dtype: int64

## 09. Correcting Bad Values

In [17]:
mapping_dict = {
    'Android': 'Android',
    'Chrome OS': 'Chrome OS',
    'Linux': 'Linux',
    'Mac OS': 'macOS',
    'No OS': 'No OS',
    'Windows': 'Windows',
    'macOS': 'macOS'
}

laptops['os'] = laptops['os'].map(mapping_dict)

In [18]:
laptops

manufacturer                                 model_name  \
0           Apple                                MacBook Pro   
1           Apple                                Macbook Air   
2              HP                                     250 G6   
3           Apple                                MacBook Pro   
4           Apple                                MacBook Pro   
5            Acer                                   Aspire 3   
6           Apple                                MacBook Pro   
7           Apple                                Macbook Air   
8            Asus                            ZenBook UX430UN   
9            Acer                                    Swift 3   
10             HP                                     250 G6   
11             HP                                     250 G6   
12          Apple                                MacBook Pro   
13           Dell                              Inspiron 3567   
14          Apple                                MacBook 12"   
15          Apple                                MacBook Pro   
16           Dell                              Inspiron 3567   
17          Apple                                MacBook Pro   
18         Lenovo                          IdeaPad 320-15IKB   
19           Dell                                     XPS 13   
20           Asus                            Vivobook E200HA   
21         Lenovo                         Legion Y520-15IKBN   
22             HP                                     255 G6   
23           Dell                              Inspiron 5379   
24             HP    15-BS101nv (i7-8550U/8GB/256GB/FHD/W10)   
25           Dell                              Inspiron 3567   
26          Apple                                MacBook Air   
27           Dell                              Inspiron 5570   
28           Dell                              Latitude 5590   
29             HP                                ProBook 470   
...           ...                                        ...   
1273           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1274         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1275         Asus                       ZenBook UX305CA-UBM1   
1276         Dell                              Inspiron 3567   
1277         Acer                             Aspire ES1-531   
1278         Dell                              Inspiron 3552   
1279       Lenovo                         IdeaPad Y700-15ISK   
1280           HP                        Pavilion 15-AW003nv   
1281         Dell                              Inspiron 3567   
1282           HP                           Stream 11-Y000na   
1283         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1284       Lenovo                             Yoga 500-14ISK   
1285       Lenovo                             Yoga 900-13ISK   
1286       Lenovo                         IdeaPad 100S-14IBR   
1287           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1288         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1289         Asus                       ZenBook UX305CA-UBM1   
1290         Dell                              Inspiron 3567   
1291         Acer                             Aspire ES1-531   
1292         Dell                              Inspiron 3552   
1293       Lenovo                         IdeaPad Y700-15ISK   
1294           HP                        Pavilion 15-AW003nv   
1295         Dell                              Inspiron 3567   
1296           HP                           Stream 11-Y000na   
1297         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1298       Lenovo                             Yoga 500-14ISK   
1299       Lenovo                             Yoga 900-13ISK   
1300       Lenovo                         IdeaPad 100S-14IBR   
1301           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1302         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   

                category screen_size

## 10. Dropping Missing Values

In [19]:
laptops_no_null_rows = laptops.dropna(axis=0)
laptops_no_null_cols = laptops.dropna(axis=1)

In [20]:
laptops_no_null_rows

manufacturer                                 model_name  \
5            Acer                                   Aspire 3   
6           Apple                                MacBook Pro   
8            Asus                            ZenBook UX430UN   
9            Acer                                    Swift 3   
13           Dell                              Inspiron 3567   
16           Dell                              Inspiron 3567   
19           Dell                                     XPS 13   
20           Asus                            Vivobook E200HA   
21         Lenovo                         Legion Y520-15IKBN   
23           Dell                              Inspiron 5379   
24             HP    15-BS101nv (i7-8550U/8GB/256GB/FHD/W10)   
25           Dell                              Inspiron 3567   
26          Apple                                MacBook Air   
27           Dell                              Inspiron 5570   
28           Dell                              Latitude 5590   
29             HP                                ProBook 470   
30          Chuwi                              LapBook 15.6"   
31           Asus       E402WA-GA010T (E2-6110/2GB/32GB/W10)   
32             HP       17-ak001nv (A6-9220/4GB/500GB/Radeon   
33           Dell                                     XPS 13   
34          Apple                                MacBook Air   
35         Lenovo                         IdeaPad 120S-14IAP   
37           Dell                              Inspiron 5770   
38             HP                                     250 G6   
39             HP                                ProBook 450   
41           Dell                              Inspiron 7577   
43           Acer                            Aspire A515-51G   
44           Dell                              Inspiron 7773   
47           Asus                                  Rog Strix   
48           Dell                              Inspiron 3567   
...           ...                                        ...   
1269         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1270       Lenovo                             Yoga 500-14ISK   
1271       Lenovo                             Yoga 900-13ISK   
1272       Lenovo                         IdeaPad 100S-14IBR   
1273           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1274         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1275         Asus                       ZenBook UX305CA-UBM1   
1276         Dell                              Inspiron 3567   
1278         Dell                              Inspiron 3552   
1279       Lenovo                         IdeaPad Y700-15ISK   
1280           HP                        Pavilion 15-AW003nv   
1282           HP                           Stream 11-Y000na   
1283         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1284       Lenovo                             Yoga 500-14ISK   
1285       Lenovo                             Yoga 900-13ISK   
1286       Lenovo                         IdeaPad 100S-14IBR   
1287           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1288         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1289         Asus                       ZenBook UX305CA-UBM1   
1290         Dell                              Inspiron 3567   
1292         Dell                              Inspiron 3552   
1293       Lenovo                         IdeaPad Y700-15ISK   
1294           HP                        Pavilion 15-AW003nv   
1296           HP                           Stream 11-Y000na   
1297         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1298       Lenovo                             Yoga 500-14ISK   
1299       Lenovo                             Yoga 900-13ISK   
1300       Lenovo                         IdeaPad 100S-14IBR   
1301           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1302         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   

                category screen_size

In [21]:
laptops_no_null_cols

manufacturer                                 model_name  \
0           Apple                                MacBook Pro   
1           Apple                                Macbook Air   
2              HP                                     250 G6   
3           Apple                                MacBook Pro   
4           Apple                                MacBook Pro   
5            Acer                                   Aspire 3   
6           Apple                                MacBook Pro   
7           Apple                                Macbook Air   
8            Asus                            ZenBook UX430UN   
9            Acer                                    Swift 3   
10             HP                                     250 G6   
11             HP                                     250 G6   
12          Apple                                MacBook Pro   
13           Dell                              Inspiron 3567   
14          Apple                                MacBook 12"   
15          Apple                                MacBook Pro   
16           Dell                              Inspiron 3567   
17          Apple                                MacBook Pro   
18         Lenovo                          IdeaPad 320-15IKB   
19           Dell                                     XPS 13   
20           Asus                            Vivobook E200HA   
21         Lenovo                         Legion Y520-15IKBN   
22             HP                                     255 G6   
23           Dell                              Inspiron 5379   
24             HP    15-BS101nv (i7-8550U/8GB/256GB/FHD/W10)   
25           Dell                              Inspiron 3567   
26          Apple                                MacBook Air   
27           Dell                              Inspiron 5570   
28           Dell                              Latitude 5590   
29             HP                                ProBook 470   
...           ...                                        ...   
1273           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1274         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1275         Asus                       ZenBook UX305CA-UBM1   
1276         Dell                              Inspiron 3567   
1277         Acer                             Aspire ES1-531   
1278         Dell                              Inspiron 3552   
1279       Lenovo                         IdeaPad Y700-15ISK   
1280           HP                        Pavilion 15-AW003nv   
1281         Dell                              Inspiron 3567   
1282           HP                           Stream 11-Y000na   
1283         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1284       Lenovo                             Yoga 500-14ISK   
1285       Lenovo                             Yoga 900-13ISK   
1286       Lenovo                         IdeaPad 100S-14IBR   
1287           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1288         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1289         Asus                       ZenBook UX305CA-UBM1   
1290         Dell                              Inspiron 3567   
1291         Acer                             Aspire ES1-531   
1292         Dell                              Inspiron 3552   
1293       Lenovo                         IdeaPad Y700-15ISK   
1294           HP                        Pavilion 15-AW003nv   
1295         Dell                              Inspiron 3567   
1296           HP                           Stream 11-Y000na   
1297         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1298       Lenovo                             Yoga 500-14ISK   
1299       Lenovo                             Yoga 900-13ISK   
1300       Lenovo                         IdeaPad 100S-14IBR   
1301           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1302         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   

                category screen_size

## 11. Filling Missing Values

In [22]:
value_counts_before = laptops.loc[laptops["os_version"].isnull(), "os"].value_counts()
laptops.loc[laptops["os"] == "macOS", "os_version"] = "X"

laptops.loc[laptops['os'] == 'No OS', 'os_version'] = 'Version Unknown'
value_counts_after = laptops.loc[laptops['os_version'].isnull(), 'os'].value_counts()

In [23]:
value_counts_after

Linux        62
Chrome OS    27
Android       2
Name: os, dtype: int64

## 12. Challenge: Clean a String Column

In [24]:
laptops['weight'] = laptops['weight'].str.replace('kgs', '').str.replace('kg', '').astype('float')

laptops.rename({'weight': 'weight_kg'}, axis=1, inplace=True)

laptops.to_csv('laptops_cleaned.csv', index=False)

In [25]:
laptops

manufacturer                                 model_name  \
0           Apple                                MacBook Pro   
1           Apple                                Macbook Air   
2              HP                                     250 G6   
3           Apple                                MacBook Pro   
4           Apple                                MacBook Pro   
5            Acer                                   Aspire 3   
6           Apple                                MacBook Pro   
7           Apple                                Macbook Air   
8            Asus                            ZenBook UX430UN   
9            Acer                                    Swift 3   
10             HP                                     250 G6   
11             HP                                     250 G6   
12          Apple                                MacBook Pro   
13           Dell                              Inspiron 3567   
14          Apple                                MacBook 12"   
15          Apple                                MacBook Pro   
16           Dell                              Inspiron 3567   
17          Apple                                MacBook Pro   
18         Lenovo                          IdeaPad 320-15IKB   
19           Dell                                     XPS 13   
20           Asus                            Vivobook E200HA   
21         Lenovo                         Legion Y520-15IKBN   
22             HP                                     255 G6   
23           Dell                              Inspiron 5379   
24             HP    15-BS101nv (i7-8550U/8GB/256GB/FHD/W10)   
25           Dell                              Inspiron 3567   
26          Apple                                MacBook Air   
27           Dell                              Inspiron 5570   
28           Dell                              Latitude 5590   
29             HP                                ProBook 470   
...           ...                                        ...   
1273           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1274         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1275         Asus                       ZenBook UX305CA-UBM1   
1276         Dell                              Inspiron 3567   
1277         Acer                             Aspire ES1-531   
1278         Dell                              Inspiron 3552   
1279       Lenovo                         IdeaPad Y700-15ISK   
1280           HP                        Pavilion 15-AW003nv   
1281         Dell                              Inspiron 3567   
1282           HP                           Stream 11-Y000na   
1283         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1284       Lenovo                             Yoga 500-14ISK   
1285       Lenovo                             Yoga 900-13ISK   
1286       Lenovo                         IdeaPad 100S-14IBR   
1287           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1288         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1289         Asus                       ZenBook UX305CA-UBM1   
1290         Dell                              Inspiron 3567   
1291         Acer                             Aspire ES1-531   
1292         Dell                              Inspiron 3552   
1293       Lenovo                         IdeaPad Y700-15ISK   
1294           HP                        Pavilion 15-AW003nv   
1295         Dell                              Inspiron 3567   
1296           HP                           Stream 11-Y000na   
1297         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1298       Lenovo                             Yoga 500-14ISK   
1299       Lenovo                             Yoga 900-13ISK   
1300       Lenovo                         IdeaPad 100S-14IBR   
1301           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1302         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   

                category screen_size

## 13. Next Steps
Our dataset is ready for some analysis, but there are still some data cleaning tasks left! Here are your next steps:

- Convert the `price_euros` column to a numeric dtype.
- Extract the screen resolution from the `screen` column.
- Extract the processor speed from the `cpu` column.

### 13.1 Convert the `price_euros` column to a numeric dtype

In [26]:
laptops['price_euros'] = laptops['price_euros'].str.replace(',', '.').astype(float)

In [27]:
laptops

manufacturer                                 model_name  \
0           Apple                                MacBook Pro   
1           Apple                                Macbook Air   
2              HP                                     250 G6   
3           Apple                                MacBook Pro   
4           Apple                                MacBook Pro   
5            Acer                                   Aspire 3   
6           Apple                                MacBook Pro   
7           Apple                                Macbook Air   
8            Asus                            ZenBook UX430UN   
9            Acer                                    Swift 3   
10             HP                                     250 G6   
11             HP                                     250 G6   
12          Apple                                MacBook Pro   
13           Dell                              Inspiron 3567   
14          Apple                                MacBook 12"   
15          Apple                                MacBook Pro   
16           Dell                              Inspiron 3567   
17          Apple                                MacBook Pro   
18         Lenovo                          IdeaPad 320-15IKB   
19           Dell                                     XPS 13   
20           Asus                            Vivobook E200HA   
21         Lenovo                         Legion Y520-15IKBN   
22             HP                                     255 G6   
23           Dell                              Inspiron 5379   
24             HP    15-BS101nv (i7-8550U/8GB/256GB/FHD/W10)   
25           Dell                              Inspiron 3567   
26          Apple                                MacBook Air   
27           Dell                              Inspiron 5570   
28           Dell                              Latitude 5590   
29             HP                                ProBook 470   
...           ...                                        ...   
1273           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1274         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1275         Asus                       ZenBook UX305CA-UBM1   
1276         Dell                              Inspiron 3567   
1277         Acer                             Aspire ES1-531   
1278         Dell                              Inspiron 3552   
1279       Lenovo                         IdeaPad Y700-15ISK   
1280           HP                        Pavilion 15-AW003nv   
1281         Dell                              Inspiron 3567   
1282           HP                           Stream 11-Y000na   
1283         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1284       Lenovo                             Yoga 500-14ISK   
1285       Lenovo                             Yoga 900-13ISK   
1286       Lenovo                         IdeaPad 100S-14IBR   
1287           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1288         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   
1289         Asus                       ZenBook UX305CA-UBM1   
1290         Dell                              Inspiron 3567   
1291         Acer                             Aspire ES1-531   
1292         Dell                              Inspiron 3552   
1293       Lenovo                         IdeaPad Y700-15ISK   
1294           HP                        Pavilion 15-AW003nv   
1295         Dell                              Inspiron 3567   
1296           HP                           Stream 11-Y000na   
1297         Asus  X556UJ-XO044T (i7-6500U/4GB/500GB/GeForce   
1298       Lenovo                             Yoga 500-14ISK   
1299       Lenovo                             Yoga 900-13ISK   
1300       Lenovo                         IdeaPad 100S-14IBR   
1301           HP        15-AC110nv (i7-6500U/6GB/1TB/Radeon   
1302         Asus        X553SA-XX031T (N3050/4GB/500GB/W10)   

                category screen_size

### 13.2 Extract the screen resolution from the `screen` column

In [28]:
screen_resolution = laptops['screen'].str.split().str[-1]

In [29]:
screen_resolution.head()

0    2560x1600
1     1440x900
2    1920x1080
3    2880x1800
4    2560x1600
Name: screen, dtype: object

### 13.3 Extract the processor speed from the `cpu` column

In [30]:
processor_speed = laptops['cpu'].str.split().str[-1]

In [31]:
processor_speed.head()

0    2.3GHz
1    1.8GHz
2    2.5GHz
3    2.7GHz
4    3.1GHz
Name: cpu, dtype: object

In [32]:
laptops.to_csv('laptops_cleaned.csv', index=False)

### 13.4 Additional Analytical Questions

Here are some questions you might like to answer in your own time by analyzing the cleaned data:

- Are laptops made by Apple more expensive than those made by other manufacturers?
- What is the best value laptop with a screen size of 15" or more?
- Which laptop has the most storage space?

The final mission in our course is a guided project, where we'll put everything together to clean and analyze a dataset using pandas!

#### 13.4.1 Are laptops made by Apple more expensive than those made by other manufacturers?

In [33]:
laptops_cleaned = pd.read_csv('laptops_cleaned.csv', encoding='Latin-1')

In [34]:
def mean_price_manufacturer(manufacturer):
    brand = laptops_cleaned[laptops_cleaned['manufacturer'] == manufacturer]
    brand_mean_price = round(brand['price_euros'].mean(), 2)
    return brand_mean_price

In [35]:
unique_manufacturer = laptops_cleaned['manufacturer'].unique()

In [36]:
print(unique_manufacturer)

['Apple' 'HP' 'Acer' 'Asus' 'Dell' 'Lenovo' 'Chuwi' 'MSI' 'Microsoft'
 'Toshiba' 'Huawei' 'Xiaomi' 'Vero' 'Razer' 'Mediacom' 'Samsung' 'Google'
 'Fujitsu' 'LG']


In [37]:
mean_price = {}
mean_price_list = []

for um in unique_manufacturer:
    mean_price[um] = mean_price_manufacturer(um)    # This function was written above 

# print(mean_price)

for key, val in mean_price.items():
    val_key = (val, key)
    mean_price_list.append(val_key)

# print(mean_price_list)

mean_price_sorted = sorted(mean_price_list, reverse=True)
# print(mean_price_sorted)

for i in mean_price_sorted:
    print(i[1], ':   ', i[0])

Razer :    3346.14
LG :    2099.0
MSI :    1728.91
Google :    1677.67
Microsoft :    1612.31
Apple :    1564.2
Huawei :    1424.0
Samsung :    1413.44
Toshiba :    1267.81
Dell :    1186.07
Xiaomi :    1133.46
Asus :    1104.17
Lenovo :    1086.38
HP :    1067.77
Fujitsu :    729.0
Acer :    626.78
Chuwi :    314.3
Mediacom :    295.0
Vero :    217.43


The mean price of several non-Apple laptops are more expensive than Apple laptops. To verify the results, we check the price of several of them.

In [38]:
razer = laptops_cleaned[laptops_cleaned['manufacturer'] == 'Razer']

razer

manufacturer     model_name   category screen_size  \
196         Razer      Blade Pro     Gaming       17.3"   
647         Razer      Blade Pro     Gaming       14.0"   
778         Razer      Blade Pro     Gaming       14.0"   
816         Razer  Blade Stealth  Ultrabook       12.5"   
830         Razer      Blade Pro     Gaming       17.3"   
1117        Razer  Blade Stealth  Ultrabook       12.5"   
1231        Razer      Blade Pro     Gaming       14.0"   

                                             screen  \
196             4K Ultra HD / Touchscreen 3840x2160   
647                               Full HD 1920x1080   
778                               Full HD 1920x1080   
816   IPS Panel 4K Ultra HD / Touchscreen 3840x2160   
830             4K Ultra HD / Touchscreen 3840x2160   
1117            Touchscreen / 4K Ultra HD 3840x2160   
1231                              Full HD 1920x1080   

                              cpu  ram_gb    storage                      gpu  \
196   Intel Core i7 7820HK 2.9GHz      32    1TB SSD  Nvidia GeForce GTX 1080   
647   Intel Core i7 7700HQ 2.8GHz      16  256GB SSD  Nvidia GeForce GTX 1060   
778   Intel Core i7 7700HQ 2.8GHz      16  512GB SSD  Nvidia GeForce GTX 1060   
816    Intel Core i7 7500U 2.5GHz      16  512GB SSD    Intel HD Graphics 620   
830   Intel Core i7 7820HK 2.9GHz      32  512GB SSD  Nvidia GeForce GTX 1080   
1117   Intel Core i7 6500U 2.5GHz       8  256GB SSD    Intel HD Graphics 520   
1231  Intel Core i7 7700HQ 2.8GHz      16    1TB SSD  Nvidia GeForce GTX 1060   

           os os_version  weight_kg  price_euros gpu_manufacturer  \
196   Windows         10       3.49       6099.0           Nvidia   
647   Windows         10       1.95       2599.0           Nvidia   
778   Windows         10       1.95       2899.0           Nvidia   
816   Windows         10       1.29       1799.0            Intel   
830   Windows         10       3.49       5499.0           Nvidia   
1117  Windows         10       1.25       1029.0            Intel   
1231  Windows         10       1.95       3499.0           Nvidia   

     cpu_manufacturer  
196             Intel  
647             Intel  
778             Intel  
816             Intel  
830             Intel  
1117            Intel  
1231            Intel

In [39]:
lg = laptops_cleaned[laptops_cleaned['manufacturer'] == 'LG']

lg

manufacturer   model_name   category screen_size  \
678           LG  Gram 15Z975  Ultrabook       15.6"   
905           LG  Gram 15Z970  Ultrabook       15.6"   
909           LG  Gram 14Z970  Ultrabook       14.0"   

                                        screen                         cpu  \
678                IPS Panel Full HD 1920x1080  Intel Core i7 8550U 1.8GHz   
905  IPS Panel Full HD / Touchscreen 1920x1080  Intel Core i7 7500U 2.7GHz   
909  IPS Panel Full HD / Touchscreen 1920x1080  Intel Core i7 7500U 2.7GHz   

     ram_gb    storage                    gpu       os os_version  weight_kg  \
678       8  512GB SSD  Intel HD Graphics 620  Windows         10       1.09   
905      16  512GB SSD  Intel HD Graphics 620  Windows         10       1.08   
909       8  512GB SSD  Intel HD Graphics 620  Windows         10       0.98   

     price_euros gpu_manufacturer cpu_manufacturer  
678       2299.0            Intel            Intel  
905       2099.0            Intel            Intel  
909       1899.0            Intel            Intel

In [40]:
msi = laptops_cleaned[laptops_cleaned['manufacturer'] == 'MSI']

msi

manufacturer                model_name category screen_size  \
58            MSI                GS73VR 7RG   Gaming       17.3"   
73            MSI                GL72M 7RDX   Gaming       17.3"   
108           MSI                GP62M 7REX   Gaming       15.6"   
121           MSI                GS63VR 7RG   Gaming       15.6"   
148           MSI               GE72MVR 7RG   Gaming       17.3"   
153           MSI                GE73VR 7RE   Gaming       17.3"   
165           MSI                GE63VR 7RE   Gaming       15.6"   
177           MSI           GT80S 6QF-074US   Gaming       18.4"   
258           MSI               GT73EVR 7RE   Gaming       17.3"   
312           MSI                GE63VR 7RF   Gaming       15.6"   
330           MSI                GL72M 7REX   Gaming       17.3"   
352           MSI              GP72MVR 7RFX   Gaming       17.3"   
393           MSI             Leopard GP72M   Gaming       17.3"   
404           MSI                GE73VR 7RF   Gaming       17.3"   
448           MSI           GV62 7RD-1686NL   Gaming       15.6"   
578           MSI              GT73VR Titan   Gaming       17.3"   
585           MSI                 GT80S 6QE   Gaming       17.3"   
603           MSI                Laptop MSI   Gaming       17.3"   
605           MSI                GS63VR 7RF   Gaming       15.6"   
618           MSI                GT62VR 7RE   Gaming       15.6"   
653           MSI                GL62M 7REX   Gaming       15.6"   
664           MSI            GP72VR Leopard   Gaming       17.3"   
675           MSI                 GP62 7RDX   Gaming       15.6"   
680           MSI                 GV62M 7RD   Gaming       15.6"   
724           MSI                GL72M 7REX   Gaming       17.3"   
738           MSI                GS63VR 7RF   Gaming       15.6"   
785           MSI                GS43VR 7RE   Gaming       14.0"   
786           MSI  GL62M (i5-7300HQ/8GB/1TB   Gaming       15.6"   
811           MSI               GE72MVR 7RG   Gaming       17.3"   
823           MSI                 GL62M 7RD   Gaming       15.6"   
894           MSI            GS73VR Stealth   Gaming       17.3"   
917           MSI                GE72VR 6RF   Gaming       17.3"   
919           MSI                 GL62M 7RD   Gaming       15.6"   
937           MSI             GP62M Leopard   Gaming       15.6"   
1032          MSI                GL72M 7RDX   Gaming       17.3"   
1047          MSI           GT72S Dominator   Gaming       17.3"   
1059          MSI                GT62VR 6RD   Gaming       15.6"   
1064          MSI                GS63VR 6RF   Gaming       15.6"   
1077          MSI          GT72VR Dominator   Gaming       17.3"   
1097          MSI              GS70 Stealth   Gaming       17.3"   
1106          MSI               GE72 Apache   Gaming       17.3"   
1138          MSI           GT72S Dominator   Gaming       17.3"   
1141          MSI                GP62M 7RDX   Gaming       15.6"   
1156          MSI                GP72M 7REX   Gaming       17.3"   
1161          MSI              GS40 Phantom   Gaming       14.0"   
1169          MSI                  PL60 7RD   Gaming       15.6"   
1178          MSI                GS60 Ghost   Gaming       15.6"   
1199          MSI                GS73VR 7RF   Gaming       17.3"   
1212          MSI            GS73VR Stealth   Gaming       17.3"   
1221          MSI               GP62MVR 6RF   Gaming       15.6"   
1229          MSI                GL62M 7RDX   Gaming       15.6"   
1233          MSI             GE72VR Apache   Gaming       17.3"   
1238          MSI                  GL62 6QF   Gaming       15.6"   
1259          MSI               GE62 Apache   Gaming       15.6"   

                               screen                          cpu  ram_gb  \
58                  Full HD 1920x1080  Intel Core i7 7700HQ 2.8GHz      16   
73                  Full HD 1920x1080  Intel Core i5 7300HQ 2.5GHz       8   
108                 Full HD 19

In [41]:
google = laptops_cleaned[laptops_cleaned['manufacturer'] == 'Google']

google

manufacturer       model_name   category screen_size  \
437       Google  Pixelbook (Core  Ultrabook       12.3"   
472       Google  Pixelbook (Core  Ultrabook       12.3"   
762       Google  Pixelbook (Core  Ultrabook       12.3"   

                    screen                        cpu  ram_gb    storage  \
437  Touchscreen 2400x1600  Intel Core i7 7Y75 1.3GHz      16  512GB SSD   
472  Touchscreen 2400x1600  Intel Core i5 7Y57 1.2GHz       8  128GB SSD   
762  Touchscreen 2400x1600  Intel Core i5 7Y57 1.2GHz       8  256GB SSD   

                       gpu         os os_version  weight_kg  price_euros  \
437  Intel HD Graphics 615  Chrome OS        NaN        1.1       2199.0   
472  Intel HD Graphics 615  Chrome OS        NaN        1.1       1275.0   
762  Intel HD Graphics 615  Chrome OS        NaN        1.1       1559.0   

    gpu_manufacturer cpu_manufacturer  
437            Intel            Intel  
472            Intel            Intel  
762            Intel            Intel

In [42]:
microsoft = laptops_cleaned[laptops_cleaned['manufacturer'] == 'Microsoft']

microsoft

manufacturer      model_name   category screen_size  \
70     Microsoft  Surface Laptop  Ultrabook       13.5"   
104    Microsoft  Surface Laptop  Ultrabook       13.5"   
449    Microsoft  Surface Laptop  Ultrabook       13.5"   
457    Microsoft  Surface Laptop  Ultrabook       13.5"   
458    Microsoft  Surface Laptop  Ultrabook       13.5"   
670    Microsoft  Surface Laptop  Ultrabook       13.5"   

                    screen                          cpu  ram_gb    storage  \
70   Touchscreen 2256x1504   Intel Core i5 7200U 2.5GHz       4  128GB SSD   
104  Touchscreen 2256x1504   Intel Core i5 7200U 2.5GHz       8  256GB SSD   
449  Touchscreen 2256x1504   Intel Core i7 7660U 2.5GHz      16  512GB SSD   
457  Touchscreen 2256x1504  Intel Core M m3-7Y30 2.2GHz       4  128GB SSD   
458  Touchscreen 2256x1504   Intel Core i7 7660U 2.5GHz       8  256GB SSD   
670  Touchscreen 2256x1504   Intel Core i7 7600U 2.8GHz       8  256GB SSD   

                              gpu       os os_version  weight_kg  price_euros  \
70          Intel HD Graphics 620  Windows       10 S      1.252      1089.00   
104         Intel HD Graphics 620  Windows       10 S      1.252      1340.00   
449  Intel Iris Plus Graphics 640  Windows       10 S      1.250      2589.00   
457         Intel HD Graphics 615  Windows       10 S      1.252       989.00   
458  Intel Iris Plus Graphics 640  Windows       10 S      1.250      1799.00   
670  Intel Iris Plus Graphics 640  Windows       10 S      1.252      1867.85   

    gpu_manufacturer cpu_manufacturer  
70             Intel            Intel  
104            Intel            Intel  
449            Intel            Intel  
457            Intel            Intel  
458            Intel            Intel  
670            Intel            Intel

In [43]:
apple = laptops_cleaned[laptops_cleaned['manufacturer'] == 'Apple']

apple

manufacturer   model_name   category screen_size  \
0           Apple  MacBook Pro  Ultrabook       13.3"   
1           Apple  Macbook Air  Ultrabook       13.3"   
3           Apple  MacBook Pro  Ultrabook       15.4"   
4           Apple  MacBook Pro  Ultrabook       13.3"   
6           Apple  MacBook Pro  Ultrabook       15.4"   
7           Apple  Macbook Air  Ultrabook       13.3"   
12          Apple  MacBook Pro  Ultrabook       15.4"   
14          Apple  MacBook 12"  Ultrabook       12.0"   
15          Apple  MacBook Pro  Ultrabook       13.3"   
17          Apple  MacBook Pro  Ultrabook       15.4"   
26          Apple  MacBook Air  Ultrabook       13.3"   
34          Apple  MacBook Air  Ultrabook       13.3"   
45          Apple  MacBook Pro  Ultrabook       13.3"   
81          Apple  MacBook 12"  Ultrabook       12.0"   
249         Apple  MacBook Pro  Ultrabook       13.3"   
270         Apple  MacBook Pro  Ultrabook       13.3"   
794         Apple  MacBook 12"  Ultrabook       12.0"   
1069        Apple  MacBook 12"  Ultrabook       12.0"   
1193        Apple  MacBook 12"  Ultrabook       12.0"   
1210        Apple  MacBook 12"  Ultrabook       12.0"   
1234        Apple  MacBook Air  Ultrabook       11.6"   

                                  screen                     cpu  ram_gb  \
0     IPS Panel Retina Display 2560x1600    Intel Core i5 2.3GHz       8   
1                               1440x900    Intel Core i5 1.8GHz       8   
3     IPS Panel Retina Display 2880x1800    Intel Core i7 2.7GHz      16   
4     IPS Panel Retina Display 2560x1600    Intel Core i5 3.1GHz       8   
6     IPS Panel Retina Display 2880x1800    Intel Core i7 2.2GHz      16   
7                               1440x900    Intel Core i5 1.8GHz       8   
12    IPS Panel Retina Display 2880x1800    Intel Core i7 2.8GHz      16   
14    IPS Panel Retina Display 2304x1440  Intel Core M m3 1.2GHz       8   
15    IPS Panel Retina Display 2560x1600    Intel Core i5 2.3GHz       8   
17    IPS Panel Retina Display 2880x1800    Intel Core i7 2.9GHz      16   
26                              1440x900    Intel Core i5 1.6GHz       8   
34                              1440x900    Intel Core i5 1.6GHz       8   
45    IPS Panel Retina Display 2560x1600    Intel Core i5 2.0GHz       8   
81    IPS Panel Retina Display 2304x1440    Intel Core i5 1.3GHz       8   
249   IPS Panel Retina Display 2560x1600    Intel Core i5 3.1GHz       8   
270   IPS Panel Retina Display 2560x1600    Intel Core i5 2.9GHz       8   
794   IPS Panel Retina Display 2304x1440     Intel Core M 1.2GHz       8   
1069  IPS Panel Retina Display 2304x1440     Intel Core M 1.1GHz       8   
1193  IPS Panel Retina Display 2304x1440     Intel Core M 1.1GHz       8   
1210  IPS Panel Retina Display 2304x1440     Intel Core M 1.2GHz       8   
1234                            1366x768    Intel Core i5 1.6GHz       4   

                  storage                           gpu     os os_version  \
0               128GB SSD  Intel Iris Plus Graphics 640  macOS          X   
1     128GB Flash Storage        Intel HD Graphics 6000  macOS          X   
3               512GB SSD            AMD Radeon Pro 455  macOS          X   
4               256GB SSD  Intel Iris Plus Graphics 650  macOS          X   
6     256GB Flash Storage       Intel Iris Pro Graphics  macOS          X   
7     256GB Flash Storage        Intel HD Graphics 6000  macOS          X   
12              256GB SSD            AMD Radeon Pro 555  macOS          X   
14              256GB SSD         Intel HD Graphics 615  macOS          X   
15              256GB SSD  Intel Iris Plus Graphics 640  macOS          X   
17              512GB SSD            AMD Radeon Pro 560  macOS          X   
26    128GB Flash Storage        Intel HD Graphics 6000  macOS          X   
34    256GB Flash Storage        Intel HD Graphics 6000  macOS          X   
45              256GB SSD       Intel Iris Graphics 540  macOS          X   

In [44]:
print('Apple — no. of entries:   ', apple.shape[0])
print('Apple — min price:        ', apple['price_euros'].min())
print('Apple — max price:        ', apple['price_euros'].max())

Apple — no. of entries:    21
Apple — min price:         898.94
Apple — max price:         2858.0


The most expensive brand is Razer, which has a wide range of laptops with the top two most expensive laptops cost 6099 EUR and 5499 EUR, respectively. 

Other non-Apple laptops that are more expensive than Apple are LG, MSI, Google and Microsoft. The laptops from these brands are within the average price range of 989 EUR to 2000++ EUR. There are only three to four entries for each of these brands, with the exception of six entries for Microsoft.

Meanwhile, there are 21 entries for Apple entries, which cover a wider price range, starting from 898.94 to 2858.00 EUR. Based on the results, it is quite convincing that laptops made by Razer are more expensive than Apple. On the other hands, although laptops made by LG, MSI, Google and Microsoft seem to be more expensive than Apple, this could also be skewed by the smaller number of entries for each of these brands. Therefore, more entries are needed for each of these brands before one can draw a solid conclusion about whether they are more expensive than Apple.

#### 13.4.2 What is the best value laptop with a screen size of 15" or more?

In [45]:
# Remove non-digit characters
laptops_cleaned['screen_size'] = laptops_cleaned['screen_size'].str.replace('"', '').astype(float)

# Group unique screen size
unique_screen_size = laptops_cleaned['screen_size'].unique()

# Add "inch" in the column label name
laptops_cleaned.rename({'screen_size': 'screen_size_inch'}, axis=1, inplace=True)

# Save the changes to the csv file
laptops_cleaned.to_csv('laptops_cleaned.csv', index=False)

In [46]:
print(unique_screen_size)

[13.3 15.6 15.4 14.  12.  11.6 17.3 10.1 13.5 12.5 13.  18.4 13.9 12.3
 17.  15.  14.1 11.3]


In [47]:
bigger_screen_laptops = laptops_cleaned[laptops_cleaned['screen_size_inch'] >= 15]

By using `nsmallest()` method, we determine the top five cheapest laptops with screen sizes of at least 15-inch as below:

In [48]:
bigger_screen_cheapest = bigger_screen_laptops.nsmallest(5, 'price_euros')

bigger_screen_cheapest

manufacturer                            model_name  category  \
290          Acer                  Chromebook C910-C2ST  Notebook   
1102         Acer                         Chromebook 15  Notebook   
555          Asus  A541NA-GO342 (N3350/4GB/500GB/Linux)  Notebook   
30          Chuwi                         LapBook 15.6"  Notebook   
483         Chuwi                          Lapbook 15,6  Notebook   

      screen_size_inch             screen  \
290               15.6           1366x768   
1102              15.6           1366x768   
555               15.6           1366x768   
30                15.6  Full HD 1920x1080   
483               15.6  Full HD 1920x1080   

                                       cpu  ram_gb             storage  \
290   Intel Celeron Dual Core 3205U 1.5GHz       2            16GB SSD   
1102  Intel Celeron Dual Core 3205U 1.5GHz       4            16GB SSD   
555   Intel Celeron Dual Core N3350 1.1GHz       4           500GB HDD   
30             Intel Atom x5-Z8300 1.44GHz       4  64GB Flash Storage   
483            Intel Atom x5-Z8350 1.44GHz       4  64GB Flash Storage   

                        gpu         os os_version  weight_kg  price_euros  \
290       Intel HD Graphics  Chrome OS        NaN       2.19       199.00   
1102      Intel HD Graphics  Chrome OS        NaN       2.20       209.00   
555   Intel HD Graphics 500      Linux        NaN       2.00       224.00   
30        Intel HD Graphics    Windows         10       1.89       244.99   
483       Intel HD Graphics    Windows         10       1.89       248.90   

     gpu_manufacturer cpu_manufacturer  
290             Intel            Intel  
1102            Intel            Intel  
555             Intel            Intel  
30              Intel            Intel  
483             Intel            Intel

Alternatively, we could also use the `sort_values()` method to determine the top five cheapest laptops with screen sizes of at least 15-inch:

In [49]:
bigger_screen_cheapest_sorted = bigger_screen_laptops.sort_values(['price_euros'])

bigger_screen_cheapest_sorted.head()

manufacturer                            model_name  category  \
290          Acer                  Chromebook C910-C2ST  Notebook   
1102         Acer                         Chromebook 15  Notebook   
555          Asus  A541NA-GO342 (N3350/4GB/500GB/Linux)  Notebook   
30          Chuwi                         LapBook 15.6"  Notebook   
483         Chuwi                          Lapbook 15,6  Notebook   

      screen_size_inch             screen  \
290               15.6           1366x768   
1102              15.6           1366x768   
555               15.6           1366x768   
30                15.6  Full HD 1920x1080   
483               15.6  Full HD 1920x1080   

                                       cpu  ram_gb             storage  \
290   Intel Celeron Dual Core 3205U 1.5GHz       2            16GB SSD   
1102  Intel Celeron Dual Core 3205U 1.5GHz       4            16GB SSD   
555   Intel Celeron Dual Core N3350 1.1GHz       4           500GB HDD   
30             Intel Atom x5-Z8300 1.44GHz       4  64GB Flash Storage   
483            Intel Atom x5-Z8350 1.44GHz       4  64GB Flash Storage   

                        gpu         os os_version  weight_kg  price_euros  \
290       Intel HD Graphics  Chrome OS        NaN       2.19       199.00   
1102      Intel HD Graphics  Chrome OS        NaN       2.20       209.00   
555   Intel HD Graphics 500      Linux        NaN       2.00       224.00   
30        Intel HD Graphics    Windows         10       1.89       244.99   
483       Intel HD Graphics    Windows         10       1.89       248.90   

     gpu_manufacturer cpu_manufacturer  
290             Intel            Intel  
1102            Intel            Intel  
555             Intel            Intel  
30              Intel            Intel  
483             Intel            Intel

In [50]:
laptops_cleaned.to_csv('laptops_cleaned.csv', index=False)

#### 13.4.2 Which laptop has the most storage space?

In [51]:
import re

# Note for Regular Expression
    # \d = Matches digits, which meansb 0-9

x = laptops_cleaned['storage'].str.extract('(\d+)').astype(float)

print(x.head())
print('\n')
print(x.tail())

       0
0  128.0
1  128.0
2  256.0
3  512.0
4  256.0


          0
1298  128.0
1299  512.0
1300   64.0
1301    1.0
1302  500.0


In [52]:
clean_storage = laptops_cleaned[x < 64]

In [53]:
clean_storage

manufacturer model_name category  screen_size_inch screen  cpu  ram_gb  \
0             NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1             NaN        NaN      NaN               NaN    NaN  NaN     NaN   
2             NaN        NaN      NaN               NaN    NaN  NaN     NaN   
3             NaN        NaN      NaN               NaN    NaN  NaN     NaN   
4             NaN        NaN      NaN               NaN    NaN  NaN     NaN   
5             NaN        NaN      NaN               NaN    NaN  NaN     NaN   
6             NaN        NaN      NaN               NaN    NaN  NaN     NaN   
7             NaN        NaN      NaN               NaN    NaN  NaN     NaN   
8             NaN        NaN      NaN               NaN    NaN  NaN     NaN   
9             NaN        NaN      NaN               NaN    NaN  NaN     NaN   
10            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
11            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
12            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
13            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
14            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
15            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
16            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
17            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
18            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
19            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
20            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
21            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
22            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
23            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
24            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
25            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
26            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
27            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
28            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
29            NaN        NaN      NaN               NaN    NaN  NaN     NaN   
...           ...        ...      ...               ...    ...  ...     ...   
1273          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1274          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1275          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1276          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1277          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1278          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1279          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1280          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1281          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1282          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1283          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1284          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1285          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1286          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1287          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1288          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1289          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1290          NaN        NaN      NaN               NaN    NaN  NaN     NaN   
1291          NaN        NaN      NaN               NaN